main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -274,59 +275,80 @@ object TypeReduction {
+    case Type.Cst(TypeConstructor.BigDecimal, _) => classOf[java.math.BigDecimal]
+    case Type.Cst(TypeConstructor.BigInt, _) => classOf[java.math.BigInteger]
+    case Type.Cst(TypeConstructor.Str, _) => classOf[String]
+    case Type.Cst(TypeConstructor.Regex, _) => classOf[java.util.regex.Pattern]
+    case Type.Cst(TypeConstructor.Native(clazz), _) => clazz
+
+    // Arrays
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Array, _), elmType, _), _, _) =>
+      // Find the Java class of the element type.
+      val t = getJavaType(elmType)
+
+      // Find the Java class of the array type.
+      // See: https://stackoverflow.com/questions/1679421/how-to-get-the-array-class-for-a-given-class-in-java
+      val phantomArr = java.lang.reflect.Array.newInstance(t, 0)
+      val arrType = phantomArr.getClass
+      arrType
+
+    // Functions
+    case Type.Apply(Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Arrow(2), _), _, _), varArg, _), varRet, _) =>
+      (varArg, varRet) match {
+        case (Type.Cst(tc1, _), Type.Cst(tc2, _)) =>
+          (tc1, tc2) match {
+            case (TypeConstructor.Int32, TypeConstructor.Unit) => classOf[java.util.function.IntConsumer]
+            case (TypeConstructor.Int32, TypeConstructor.Bool) => classOf[java.util.function.IntPredicate]
+            case (TypeConstructor.Int32, TypeConstructor.Int32) => classOf[java.util.function.IntUnaryOperator]
+            case (TypeConstructor.Int32, TypeConstructor.Native(obj)) if obj == classOf[Object] => classOf[java.util.function.IntFunction[Object]]
+            case (TypeConstructor.Float64, TypeConstructor.Unit) => classOf[java.util.function.DoubleConsumer]
+            case (TypeConstructor.Float64, TypeConstructor.Bool) => classOf[java.util.function.DoublePredicate]
+            case (TypeConstructor.Float64, TypeConstructor.Float64) => classOf[java.util.function.DoubleUnaryOperator]
+            case (TypeConstructor.Float64, TypeConstructor.Native(obj)) if obj == classOf[Object] => classOf[java.util.function.DoubleFunction[Object]]
+            case (TypeConstructor.Int64, TypeConstructor.Unit) => classOf[java.util.function.LongConsumer]
+            case (TypeConstructor.Int64, TypeConstructor.Bool) => classOf[java.util.function.LongPredicate]
+            case (TypeConstructor.Int64, TypeConstructor.Int64) => classOf[java.util.function.LongUnaryOperator]
+            case (TypeConstructor.Int64, TypeConstructor.Native(obj)) if obj == classOf[Object] => classOf[java.util.function.LongFunction[Object]]
+            case (TypeConstructor.Native(obj), TypeConstructor.Unit) if obj == classOf[Object] => classOf[java.util.function.Consumer[Object]]
+            case (TypeConstructor.Native(obj), TypeConstructor.Bool) if obj == classOf[Object] => classOf[java.util.function.Predicate[Object]]
+            case _ => classOf[Object] // default
+          }
+        case _ => classOf[Object] // default
+      }
+    case _ => classOf[Object] // default
```

Comment:
Returning object should be OK: We know its not a primitive, so it must be assigned to object.

The alternative is to write all cases out. We could also do that-- then we need cases for tuples, vectors, bla bla.

Reply:
hmm. i guess its fine yeah